firebase-segmentation/src/main/java/com/google/firebase/segmentation/CustomInstallationIdCache.java

```diff
@@ -0,0 +1,109 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.firebase.segmentation;
+
+import android.content.SharedPreferences;
+import androidx.annotation.Nullable;
+import androidx.annotation.VisibleForTesting;
+import com.google.android.gms.common.util.Strings;
+import com.google.firebase.FirebaseApp;
+
+class CustomInstallationIdCache {
+
+  // Status of each cache entry
+  // NOTE: never change the ordinal of the enum values because the enum values are stored in cache
+  // as their ordinal numbers.
+  enum CacheStatus {
+    // Cache entry is synced to Firebase backend
+    SYNCED,
+    // Cache entry is waiting for Firebase backend response or pending internal retry for retryable
+    // errors.
+    PENDING,
+    // Cache entry is not accepted by Firebase backend.
+    ERROR,
+  }
+
+  private static final String SHARED_PREFS_NAME = "CustomInstallationIdCache";
+
+  private static final String CUSTOM_INSTALLATION_ID_KEY = "Cid";
+  private static final String INSTANCE_ID_KEY = "Iid";
+  private static final String CACHE_STATUS_KEY = "Status";
+
+  private static CustomInstallationIdCache singleton = null;
+  private final SharedPreferences prefs;
+
+  static CustomInstallationIdCache getInstance() {
```

Comment:
Due to the way SharedPrefs are implemented, this class does not have to be a singleton, as multiple instances of sharedPrefs that point to the same file share the same thread-safe in-memory backing structure. As a side-effect of doing it your interface will become much simpler, as you won't have to pass FirebaseApp around to be able to read/write to the cache, e.g. something along these lines:

```java
class Cache {
  // note: injecting direct deps only https://github.com/google/guice/wiki/InjectOnlyDirectDependencies
  Cache(String persistenceKey, SharedPreferences prefs) {}
  Value read();
  void write(Value value);
}
```

When your SDK initializes it can then construct this cache as `new Cache(app.getPersistenceKey(), context.getSharedPreferences());` Note that Cache is completely decoupled from FirebaseApp, so there is not need to initialize FirebaseApp during tests, which leads to faster more isolated tests. wdyt?

Reply:
I agree with the multi cache instance idea.

But for the constructor arguments, I still prefer Cache(FirebaseApp app),
I don't want the caller to pass in a random string other than a app.getPersistenceKey().
Also the shared pref, I want to hide the shared pref name inside of the cache class.

Initializing FirebaseApp in tests doesn't hurt, which is fine, I think

Reply:
1. For passing in a String argument and SharedPreference argument from the caller, I don't like the idea, because I don't want the caller to pass in a random string or a random shared pref name.

2. For singleton or non-singleton, I don't have a strong opinion. But the singleton mode can centralize management of all FirebaseApps in one shared pref file, which allows to clearAll() for example, and it doesn't hurt the test too much.

Reply:
I don't feel strongly re 1.

2. you can centralize all FirebaseApps in one shared pref file even if you don't use a singleton, as calling `getSharedPreferences()` with the same name+mode returns the same instance, clearAll can also be implemented as a static method that takes a `Context`. It's ultimately up to your team to decide the best approach here, but I feel like singleton does not provide many benefits due to the nature of how sharedPrefs work. If you do decide to use a singleton, pls make `getInstance()` thread-safe.

Reply:
Changing to thread-safe singleton.

If we find singleton pattern doesn't work well later, we can always change it.